### PR TITLE
`ControllerName` validation helper

### LINKED ISSUE #1410: `ControllerName` validation helper
**What would you like to be added**:

The GatewayClass `ControllerName` validation is enforced at the CRD level through the following regex:
`^[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*\/[A-Za-z0-9\/\-._~%!$&'()*+,;=:]+$`

Since the controllerName is a field that may be used by the implementations to check that the `GatewayClass` has to be reconciled by the controller itself, it would be beneficial to provide a helper for checking its validity. This helper could be added under `apis/v1beta1/validation` or `apis/v1beta1/util/validation`

**Why this is needed**:
This is needed to provide a helper to check the `ControllerName` validity by the various implementations.

### PR DESCRIPTION
**Pull request by mlavacca**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:
A utility to check the Controller name validity has been added to apis/<version>/util/validation.

**Which issue(s) this PR fixes**:

Fixes #1410 

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: apis/v1alpha2/util/validation/gatewayclass.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+package validation
+
+import (
+	"regexp"
+
+	gatewayv1a2 "sigs.k8s.io/gateway-api/apis/v1alpha2"
+)
+
+// IsControllerNameValid checks that the provided controllerName complaints with the expected
```
Comment (sunjayBhatia) [inline apis/v1alpha2/util/validation/gatewayclass.go:25]: spelling nit, same for the other copy of this

```suggestion
// IsControllerNameValid checks that the provided controllerName complies with the expected
```

File: apis/v1beta1/util/validation/gatewayclass.go [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+package validation
+
+import (
+	"regexp"
+
+	gatewayv1b1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+)
+
+// IsControllerNameValid checks that the provided controllerName complaints with the expected
```
Comment (akankshakumari393) [inline apis/v1beta1/util/validation/gatewayclass.go:25]:
```suggestion
// IsControllerNameValid checks that the provided controllerName complies with the expected
```

File: apis/v1alpha2/util/validation/gatewayclass_test.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+package validation_test
+
+import (
+	"testing"
+
+	gatewayv1a2 "sigs.k8s.io/gateway-api/apis/v1alpha2"
+	validationtutils "sigs.k8s.io/gateway-api/apis/v1alpha2/util/validation"
+)
+
+func TestIsControllerNameValid(t *testing.T) {
```
Comment (sunjayBhatia) [inline apis/v1alpha2/util/validation/gatewayclass_test.go:26]: maybe worth adding another test or two that exercise the regex more, w/ invalid characters etc.

**Review by howardjohn — COMMENTED** [review]
TBH I don't get the use case. Seems like each impl will make pick `controllerName` as a constant, and if its invalid they will pick a new one? Not sure why we need a Go helper...

I also don't mind though

File: apis/v1alpha2/util/validation/gatewayclass_test.go (hunk 3 above)
Comment (mlavacca) [reply apis/v1alpha2/util/validation/gatewayclass_test.go:26]: I added a couple more test cases.

File: apis/v1alpha2/util/validation/gatewayclass.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+// format. It must be a non-empty domain prefixed path.
+func IsControllerNameValid(controllerName gatewayv1a2.GatewayController) bool {
+	if controllerName == "" {
+		return false
+	}
+
+	re := regexp.MustCompile(`^[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*\/[A-Za-z0-9\/\-._~%!$&'()*+,;=:]+$`)
```
[the remark this answers is not in the document]
Comment (mlavacca) [reply apis/v1alpha2/util/validation/gatewayclass.go]: I've set it as a global variable so that it is compiled only once.

**Comment by mlavacca** [comment]
> TBH I don't get the use case. Seems like each impl will make pick `controllerName` as a constant, and if its invalid they will pick a new one? Not sure why we need a Go helper...

Recently in the Kong implementation, we have had to make it customizable via argument. This helper is meant to validate the controllerName in such a use case.

**Comment by akankshakumari393** [comment]
Yeah it happens, If you request one, the others would be removed from review request.

**Comment by youngnick** [comment]
For the case that people take the controllerName as an argument, having common code to validate seems like a great idea to me.

/approve

I'll leave the LGTM for someone else though.

**Review by shaneutt — CHANGES_REQUESTED** [review]
Code and idea LGTM, however one thing I would like to talk about possibly changing:

We're putting this in two places with the same regex. Having copied code is usually not ideal. We recently made changes to share types between v1alpha2 and v1beta1 to reduce some of these copies, I think we should do the same here.

**Comment by mlavacca** [comment]
> We're putting this in two places with the same regex. Having copied code is usually not ideal. We recently made changes to share types between v1alpha2 and v1beta1 to reduce some of these copies, I think we should do the same here.

I factorized the regexp and kept only the `v1beta1` instance.

**Review by shaneutt — COMMENTED** [review]
We should be able to share _all_ the v1beta1 code backwards to v1alpha2, rather than just the regexp no?

**Pull request merged**